ign_ros2_control/src/ign_system.cpp

```diff
@@ -580,42 +776,133 @@ IgnitionSystem::perform_command_mode_switch(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdMin_vel);
+    this->dataPtr->joints_[i].pid_vel.SetCmdOffset(
+      params_.gains.joints_map[this->dataPtr->joints_[i].
+      name].cmdOffset_vel);
+
     if (this->dataPtr->joints_[i].joint_control_method & VELOCITY) {
-      if (!this->dataPtr->ecm->Component<ignition::gazebo::components::JointVelocityCmd>(
-          this->dataPtr->joints_[i].sim_joint))
-      {
+
+      double velocity_error = this->dataPtr->joints_[i].joint_velocity - std::clamp(
+        this->dataPtr->joints_[i].joint_velocity_cmd, -1.0 * jointAxis->Data().MaxVelocity(),
+        jointAxis->Data().MaxVelocity());
+
+      // calculate target force/torque - output of inner pid
+      double target_force = this->dataPtr->joints_[i].pid_vel.Update(
+        velocity_error, std::chrono::duration<double>(
+          period.to_chrono<std::chrono::nanoseconds>()));
+
+      // remember for potential effort state interface
+      this->dataPtr->joints_[i].joint_effort_cmd = target_force;
+
+      auto forceCmd =
+        this->dataPtr->ecm->Component<ignition::gazebo::components::JointForceCmd>(
+        this->dataPtr->joints_[i].sim_joint);
+
+      if (forceCmd == nullptr) {
         this->dataPtr->ecm->CreateComponent(
           this->dataPtr->joints_[i].sim_joint,
-          ignition::gazebo::components::JointVelocityCmd({0}));
+          ignition::gazebo::components::JointForceCmd({target_force}));
       } else {
-        const auto jointVelCmd =
-          this->dataPtr->ecm->Component<ignition::gazebo::components::JointVelocityCmd>(
-          this->dataPtr->joints_[i].sim_joint);
-        *jointVelCmd = ignition::gazebo::components::JointVelocityCmd(
-          {this->dataPtr->joints_[i].joint_velocity_cmd});
+        *forceCmd = ignition::gazebo::components::JointForceCmd(
+          {target_force});
       }
-    } else if (this->dataPtr->joints_[i].joint_control_method & POSITION) {
-      // Get error in position
-      double error;
-      error = (this->dataPtr->joints_[i].joint_position -
-        this->dataPtr->joints_[i].joint_position_cmd) * *this->dataPtr->update_rate;
 
-      // Calculate target velcity
-      double target_vel = -this->dataPtr->position_proportional_gain_ * error;
+    } else if (this->dataPtr->joints_[i].joint_control_method & POSITION) {
 
-      auto vel =
-        this->dataPtr->ecm->Component<ignition::gazebo::components::JointVelocityCmd>(
+      // calculate error with clamped position command
+      double position_error = this->dataPtr->joints_[i].joint_position - std::clamp(
+        this->dataPtr->joints_[i].joint_position_cmd, jointAxis->Data().Lower(),
+        jointAxis->Data().Upper());
+      position_error =
+        copysign(
+        1.0,
+        position_error) *
+        std::clamp(
+        std::abs(position_error), 0.0,
+        std::abs(jointAxis->Data().Upper() - jointAxis->Data().Lower()));
```

Comment:
Possible to split this up? :thinking:

Reply:
[e40142c](https://github.com/ros-controls/gz_ros2_control/pull/122/commits/e40142c66370bbab409963333b429863e6dfabe1)